File: aws_hit_breaks/services/rds.py
"""
RDS service manager for discovering and managing RDS instances and clusters.
"""
from typing import List, Dict, Any
from datetime import datetime

from .base import BaseServiceManager
from .models import Resource, OperationResult
from ..core.exceptions import ServiceError


class RDSServiceManager(BaseServiceManager):
    """Service manager for RDS instances and Aurora clusters."""
    
    @property
    def service_name(self) -> str:
        return 'rds'
    
    def discover_resources(self) -> List[Resource]:
        """Discover all RDS instances and Aurora clusters in the region.
        
        Returns:
            List of RDS resources as Resource objects
            
        Raises:
            ServiceError: If discovery fails
        """
        try:
            resources = []
            
            # Discover RDS instances
            db_instances = self.client.describe_db_instances()
            for instance in db_instances['DBInstances']:
                # Skip instances that are being deleted
                if instance['DBInstanceStatus'] == 'deleting':
                    continue
                
                # Extract tags
                tags = {}
                try:
                    tag_response = self.client.list_tags_for_resource(
                        ResourceName=instance['DBInstanceArn']
                    )
                    for tag in tag_response['TagList']:
                        tags[tag['Key']] = tag['Value']
                except Exception:
                    # Tags might not be accessible, continue without them
                    pass
                
                resource = Resource(
                    service_type='rds',
                    resource_id=instance['DBInstanceIdentifier'],
                    region=self.region,
                    current_state=instance['DBInstanceStatus'],
                    tags=tags,
                    metadata={
                        'engine': instance['Engine'],
                        'engine_version': instance['EngineVersion'],
                        'instance_class': instance['DBInstanceClass'],
                        'allocated_storage': instance.get('AllocatedStorage'),
                        'storage_type': instance.get('StorageType'),
                        'multi_az': instance.get('MultiAZ', False),
                        'availability_zone': instance.get('AvailabilityZone'),
                        'vpc_security_groups': [sg['VpcSecurityGroupId'] for sg in instance.get('VpcSecurityGroups', [])],
                        'db_subnet_group': instance.get('DBSubnetGroup', {}).get('DBSubnetGroupName'),
                        'resource_type': 'db_instance'
                    }
                )
                resources.append(resource)
            
            # Discover Aurora clusters
            db_clusters = self.client.describe_db_clusters()
            for cluster in db_clusters['DBClusters']:
                # Skip clusters that are being deleted
                if cluster['Status'] == 'deleting':
                    continue
                
                # Extract tags
                tags = {}
                try:
                    tag_response = self.client.list_tags_for_resource(
                        ResourceName=cluster['DBClusterArn']
                    )
                    for tag in tag_response['TagList']:
                        tags[tag['Key']] = tag['Value']
                except Exception:
                    # Tags might not be accessible, continue without them
                    pass
                
                resource = Resource(
                    service_type='rds',
                    resource_id=cluster['DBClusterIdentifier'],
                    region=self.region,
                    current_state=cluster['Status'],
                    tags=tags,
                    metadata={
                        'engine': cluster['Engine'],
                        'engine_version': cluster['EngineVersion'],
                        'cluster_members': [member['DBInstanceIdentifier'] for member in cluster.get('DBClusterMembers', [])],
                        'multi_az': cluster.get('MultiAZ', False),
                        'availability_zones': cluster.get('AvailabilityZones', []),
                        'vpc_security_groups': [sg['VpcSecurityGroupId'] for sg in cluster.get('VpcSecurityGroups', [])],
                        'db_subnet_group': cluster.get('DBSubnetGroup'),
                        'resource_type': 'db_cluster'
                    }
                )
                resources.append(resource)
            
            return resources
            
        except Exception as e:
            self._handle_aws_error(e, 'discovery')
    
    def pause_resource(self, resource: Resource) -> OperationResult:
        """Stop an RDS instance or cluster.
        
        Args:
            resource: RDS resource to stop
            
        Returns:
            Result of the stop operation
        """
        start_time = datetime.now()
        
        try:
            resource_type = resource.metadata.get('resource_type')
            
            if resource_type == 'db_instance':
                return self._pause_db_instance(resource, start_time)
            elif resource_type == 'db_cluster':
                return self._pause_db_cluster(resource, start_time)
            else:
                return self._create_operation_result(
                    resource=resource,
                    operation='pause',
                    success=False,
                    message=f"Unknown RDS resource type: {resource_type}",
                    start_time=start_time,
                    duration=0.0
                )
                
        except Exception as e:
            duration = (datetime.now() - start_time).total_seconds()
            return self._create_operation_result(
                resource=resource,
                operation='pause',
                success=False,
                message=f"Failed to stop RDS resource {resource.resource_id}: {str(e)}",
                start_time=start_time,
                duration=duration
            )
    
    def resume_resource(self, resource: Resource) -> OperationResult:
        """Start an RDS instance or cluster.
        
        Args:
            resource: RDS resource to start
            
        Returns:
            Result of the start operation
        """
        start_time = datetime.now()
        
        try:
            resource_type = resource.metadata.get('resource_type')
            
            if resource_type == 'db_instance':
                return self._resume_db_instance(resource, start_time)
            elif resource_type == 'db_cluster':
                return self._resume_db_cluster(resource, start_time)
            else:
                return self._create_operation_result(
                    resource=resource,
                    operation='resume',
                    success=False,
                    message=f"Unknown RDS resource type: {resource_type}",
                    start_time=start_time,
                    duration=0.0
                )
                
        except Exception as e:
            duration = (datetime.now() - start_time).total_seconds()
            return self._create_operation_result(
                resource=resource,
                operation='resume',
                success=False,
                message=f"Failed to start RDS resource {resource.resource_id}: {str(e)}",
                start_time=start_time,
                duration=duration
            )
    
    def _pause_db_instance(self, resource: Resource, start_time: datetime) -> OperationResult:
        """Stop a DB instance."""
        # Check if instance can be stopped
        if resource.current_state not in ['available']:
            return self._create_operation_result(
                resource=resource,
                operation='pause',
                success=False,
                message=f"DB instance {resource.resource_id} cannot be stopped (current state: {resource.current_state})",
                start_time=start_time,
                duration=0.0
            )
        
        # Stop the instance
        self.client.stop_db_instance(DBInstanceIdentifier=resource.resource_id)
        
        # Wait for instance to stop
        waiter = self.client.get_waiter('db_instance_stopped')
        waiter.wait(
            DBInstanceIdentifier=resource.resource_id,
            WaiterConfig={'Delay': 30, 'MaxAttempts': 60}  # Wait up to 30 minutes
        )
        
        duration = (datetime.now() - start_time).total_seconds()
        
        return self._create_operation_result(
            resource=resource,
            operation='pause',
            success=True,
            message=f"Successfully stopped RDS instance {resource.resource_id}",
            start_time=start_time,
            duration=duration
        )
    
    def _resume_db_instance(self, resource: Resource, start_time: datetime) -> OperationResult:
        """Start a DB instance."""
        # Check if instance can be started
        if resource.current_state not in ['stopped']:
            return self._create_operation_result(
                resource=resource,
                operation='resume',
                success=False,
                message=f"DB instance {resource.resource_id} cannot be started (current state: {resource.current_state})",
                start_time=start_time,
                duration=0.0
            )
        
        # Start the instance
        self.client.start_db_instance(DBInstanceIdentifier=resource.resource_id)
        
        # Wait for instance to be available
        waiter = self.client.get_waiter('db_instance_available')
        waiter.wait(
            DBInstanceIdentifier=resource.resource_id,
            WaiterConfig={'Delay': 30, 'MaxAttempts': 60}  # Wait up to 30 minutes
        )
        
        duration = (datetime.now() - start_time).total_seconds()
        
        return self._create_operation_result(
            resource=resource,
            operation='resume',
            success=True,
            message=f"Successfully started RDS instance {resource.resource_id}",
            start_time=start_time,
            duration=duration
        )
    
    def _pause_db_cluster(self, resource: Resource, start_time: datetime) -> OperationResult:
        """Stop a DB cluster."""
        # Check if cluster can be stopped
        if resource.current_state not in ['available']:
            return self._create_operation_result(
                resource=resource,
                operation='pause',
                success=False,
                message=f"DB cluster {resource.resource_id} cannot be stopped (current state: {resource.current_state})",
                start_time=start_time,
                duration=0.0
            )
        
        # Stop the cluster
        self.client.stop_db_cluster(DBClusterIdentifier=resource.resource_id)
        
        # Wait for cluster to stop
        waiter = self.client.get_waiter('db_cluster_stopped')
        waiter.wait(
            DBClusterIdentifier=resource.resource_id,
            WaiterConfig={'Delay': 30, 'MaxAttempts': 60}  # Wait up to 30 minutes
        )
        
        duration = (datetime.now() - start_time).total_seconds()
        
        return self._create_operation_result(
            resource=resource,
            operation='pause',
            success=True,
            message=f"Successfully stopped RDS cluster {resource.resource_id}",
            start_time=start_time,
            duration=duration
        )
    
    def _resume_db_cluster(self, resource: Resource, start_time: datetime) -> OperationResult:
        """Start a DB cluster."""
        # Check if cluster can be started
        if resource.current_state not in ['stopped']:
            return self._create_operation_result(
                resource=resource,
                operation='resume',
                success=False,
                message=f"DB cluster {resource.resource_id} cannot be started (current state: {resource.current_state})",
                start_time=start_time,
                duration=0.0
            )
        
        # Start the cluster
        self.client.start_db_cluster(DBClusterIdentifier=resource.resource_id)
        
        # Wait for cluster to be available
        waiter = self.client.get_waiter('db_cluster_available')
        waiter.wait(
            DBClusterIdentifier=resource.resource_id,
            WaiterConfig={'Delay': 30, 'MaxAttempts': 60}  # Wait up to 30 minutes
        )
        
        duration = (datetime.now() - start_time).total_seconds()
        
        return self._create_operation_result(
            resource=resource,
            operation='resume',
            success=True,
            message=f"Successfully started RDS cluster {resource.resource_id}",
            start_time=start_time,
            duration=duration
        )

File: internal/services/rds.go
package services

import (
	"context"
	"fmt"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/rds"
	"github.com/aws/aws-sdk-go-v2/service/rds/types"

	"github.com/aicoder2009/aws-hit-breaks/internal/models"
)

// RDSServiceManager handles RDS instance and cluster operations
type RDSServiceManager struct {
	client *rds.Client
	region string
}

// NewRDSServiceManager creates a new RDS service manager
func NewRDSServiceManager(cfg aws.Config) *RDSServiceManager {
	return &RDSServiceManager{
		client: rds.NewFromConfig(cfg),
		region: cfg.Region,
	}
}

// ServiceType returns the service type
func (m *RDSServiceManager) ServiceType() models.ServiceType {
	return models.ServiceRDS
}

// Discover finds all running RDS instances and Aurora clusters
func (m *RDSServiceManager) Discover(ctx context.Context, region string) ([]models.Resource, error) {
	var resources []models.Resource

	// Discover RDS instances
	instances, err := m.discoverInstances(ctx, region)
	if err != nil {
		return nil, err
	}
	resources = append(resources, instances...)

	// Discover Aurora clusters
	clusters, err := m.discoverClusters(ctx, region)
	if err != nil {
		return nil, err
	}
	resources = append(resources, clusters...)

	return resources, nil
}

func (m *RDSServiceManager) discoverInstances(ctx context.Context, region string) ([]models.Resource, error) {
	var resources []models.Resource

	paginator := rds.NewDescribeDBInstancesPaginator(m.client, &rds.DescribeDBInstancesInput{})
	for paginator.HasMorePages() {
		output, err := paginator.NextPage(ctx)
		if err != nil {
			return nil, fmt.Errorf("failed to describe RDS instances: %w", err)
		}

		for _, instance := range output.DBInstances {
			// Skip instances that are part of Aurora clusters (handled separately)
			if instance.DBClusterIdentifier != nil {
				continue
			}

			// Only include available (running) instances
			if aws.ToString(instance.DBInstanceStatus) != "available" {
				continue
			}

			resource := m.instanceToResource(instance, region)
			resources = append(resources, resource)
		}
	}

	return resources, nil
}

func (m *RDSServiceManager) discoverClusters(ctx context.Context, region string) ([]models.Resource, error) {
	var resources []models.Resource

	paginator := rds.NewDescribeDBClustersPaginator(m.client, &rds.DescribeDBClustersInput{})
	for paginator.HasMorePages() {
		output, err := paginator.NextPage(ctx)
		if err != nil {
			return nil, fmt.Errorf("failed to describe RDS clusters: %w", err)
		}

		for _, cluster := range output.DBClusters {
			// Only include available clusters
			if aws.ToString(cluster.Status) != "available" {
				continue
			}

			resource := m.clusterToResource(cluster, region)
			resources = append(resources, resource)
		}
	}

	return resources, nil
}

// Pause stops an RDS instance or cluster
func (m *RDSServiceManager) Pause(ctx context.Context, resource models.Resource) error {
	isCluster := resource.Metadata["is_cluster"] == true

	if isCluster {
		_, err := m.client.StopDBCluster(ctx, &rds.StopDBClusterInput{
			DBClusterIdentifier: aws.String(resource.ResourceID),
		})
		if err != nil {
			return fmt.Errorf("failed to stop RDS cluster %s: %w", resource.ResourceID, err)
		}
	} else {
		_, err := m.client.StopDBInstance(ctx, &rds.StopDBInstanceInput{
			DBInstanceIdentifier: aws.String(resource.ResourceID),
		})
		if err != nil {
			return fmt.Errorf("failed to stop RDS instance %s: %w", resource.ResourceID, err)
		}
	}

	return nil
}

// Resume starts an RDS instance or cluster
func (m *RDSServiceManager) Resume(ctx context.Context, resource models.Resource) error {
	isCluster := resource.Metadata["is_cluster"] == true

	if isCluster {
		_, err := m.client.StartDBCluster(ctx, &rds.StartDBClusterInput{
			DBClusterIdentifier: aws.String(resource.ResourceID),
		})
		if err != nil {
			return fmt.Errorf("failed to start RDS cluster %s: %w", resource.ResourceID, err)
		}
	} else {
		_, err := m.client.StartDBInstance(ctx, &rds.StartDBInstanceInput{
			DBInstanceIdentifier: aws.String(resource.ResourceID),
		})
		if err != nil {
			return fmt.Errorf("failed to start RDS instance %s: %w", resource.ResourceID, err)
		}
	}

	return nil
}

func (m *RDSServiceManager) instanceToResource(instance types.DBInstance, region string) models.Resource {
	// Extract tags
	tags := make(map[string]string)
	for _, tag := range instance.TagList {
		if tag.Key != nil && tag.Value != nil {
			tags[*tag.Key] = *tag.Value
		}
	}

	metadata := map[string]any{
		"is_cluster":     false,
		"engine":         aws.ToString(instance.Engine),
		"engine_version": aws.ToString(instance.EngineVersion),
		"instance_class": aws.ToString(instance.DBInstanceClass),
		"multi_az":       instance.MultiAZ,
	}

	if instance.AllocatedStorage != nil {
		metadata["storage_gb"] = *instance.AllocatedStorage
	}

	costPerHour := estimateRDSCost(aws.ToString(instance.DBInstanceClass), aws.ToString(instance.Engine), region)

	return models.Resource{
		ServiceType:  models.ServiceRDS,
		ResourceID:   aws.ToString(instance.DBInstanceIdentifier),
		Region:       region,
		CurrentState: models.StateAvailable,
		Tags:         tags,
		Metadata:     metadata,
		CostPerHour:  costPerHour,
	}
}

func (m *RDSServiceManager) clusterToResource(cluster types.DBCluster, region string) models.Resource {
	// Extract tags
	tags := make(map[string]string)
	for _, tag := range cluster.TagList {
		if tag.Key != nil && tag.Value != nil {
			tags[*tag.Key] = *tag.Value
		}
	}

	metadata := map[string]any{
		"is_cluster":     true,
		"engine":         aws.ToString(cluster.Engine),
		"engine_version": aws.ToString(cluster.EngineVersion),
	}

	if cluster.AllocatedStorage != nil {
		metadata["storage_gb"] = *cluster.AllocatedStorage
	}

	return models.Resource{
		ServiceType:  models.ServiceRDS,
		ResourceID:   aws.ToString(cluster.DBClusterIdentifier),
		Region:       region,
		CurrentState: models.StateAvailable,
		Tags:         tags,
		Metadata:     metadata,
		CostPerHour:  0.10, // Aurora cluster base cost
	}
}

func estimateRDSCost(instanceClass, engine, region string) float64 {
	// Simplified pricing
	pricing := map[string]float64{
		"db.t3.micro":  0.017,
		"db.t3.small":  0.034,
		"db.t3.medium": 0.068,
		"db.t3.large":  0.136,
		"db.m5.large":  0.171,
		"db.m5.xlarge": 0.342,
		"db.r5.large":  0.24,
		"db.r5.xlarge": 0.48,
	}

	if cost, ok := pricing[instanceClass]; ok {
		return cost
	}
	return 0.10 // Default estimate
}
